rust/noosphere-gateway/src/again.rs

```diff
@@ -0,0 +1,147 @@
+use anyhow::Result;
+use std::future::Future;
+use std::sync::Arc;
+use tokio::sync::OnceCell;
+
+/// Wraps an "initialization" step, expressed as a closure, and allows
+/// a user to invoke closures with the result of that initialization step
+/// as a context argument. The result of invocation is always returned, but
+/// a failure result causes the initialized context to be reset is that it
+/// is re-initialized upon the next invocation attempt.
+pub struct Again<I, O, F>
+where
+    F: Future<Output = Result<O, anyhow::Error>>,
+    I: Fn() -> F,
+{
+    init: I,
+    initialized: OnceCell<Arc<O>>,
+}
+
+impl<I, O, F> Again<I, O, F>
+where
+    F: Future<Output = Result<O, anyhow::Error>>,
+    I: Fn() -> F,
+{
+    pub fn new(init: I) -> Self {
+        Again {
+            init,
+            initialized: OnceCell::new(),
+        }
+    }
+
+    /// Invoke a closure with the initialized context. The result will be
+    /// returned as normal, but an error result will cause the initialized
+    /// context to be reset so that the next time an invocation is attempted,
+    /// context initialization will be retried.
+    pub async fn invoke_or_reset<Ii, Oo, Ff>(&mut self, invoke: Ii) -> Result<Oo>
```

Comment:
This function either invokes AND resets, or just invokes -- possibly rename to just `invoke`, as this whole struct/state is the "reset"